### DISCUSSION TITLE
C unit testing with multiple test files

### DISCUSSION THREAD

**Discussion by student868** [google/googletest#3605] [Community Help]
Hi,

I am trying to use multiple test files for C code.
I include the same headers on both files inside extern block but keep getting LNK2005 error (on one file I include also the .c files because it won't compile otherwise).

can someone have an idea why? (I do have guards inside the header files).

Thanks

**Comment by joshiayush** [comment]
# Linker Tools Error LNK2005

> This error can occur when a header file defines a function that isn't inline . If you include this header file in more than one source file, you get multiple definitions of the function in the executable.

Or in general if you define your functions in _header files_ and include them in multiple _object files_ then you're gonna get this error anyways.

Could you please post the code that is creating that error?

**Reply by student868** [reply]
hi,

thanks for your reply. I have the following files:

try.h
```
#ifndef TRY_TRY_H
#define TRY_TRY_H

#include <stdio.h>

int hello_world();

#endif //TRY_TRY_H
```

try.c
```
#include "try.h"

int hello_world(){
    printf("hello world!\n");
    return 0;
}
```

tests.cpp
```
#include <fstream>
#include "gtest/gtest.h"

extern "C" {
#include "../../try.h"
#include "../../try.c"
}

TEST(t1, t1a) {
    EXPECT_EQ(hello_world(), 0);
}
```

tests2.cpp
```
#include <fstream>
#include "gtest/gtest.h"

extern "C" {
#include "../../try.h"
#include "../../try.c"
}

TEST(t2, t2a) {
    EXPECT_EQ(hello_world(), 0);
}
```

CMakeLists.txt:
```
project(Google_tests)

add_subdirectory(lib)
include_directories(${gtest_SOURCE_DIR}/include ${gtest_SOURCE_DIR})

add_executable(Google_Tests_run tests/tests.cpp tests/tests2.cpp)

target_link_libraries(Google_Tests_run gtest gtest_main)
```

How can I use ".c" & ".h" files in multiple test files with google test?

thanks

**Reply by ghost** [reply]
@student868 

I think it seems also good to have multiple executables(test binaries) rather than having just one executable
And then you can run all of them at once with the command `ctest` (or `make test` or `ninja test`) on shell

```cmake
enable_testing()

function(create_test name)
  add_executable(${name} tests/${name}.cpp)
  target_link_libraries(${name} gtest gtest_main)
  add_test(NAME ${name} COMMAND "$<TARGET_FILE:${name}>")
endfunction()

create_test(tests)    # for tests/tests.cpp
create_test(tests2)   # for tests/tests2.cpp
```
Another practice can be found here : https://cliutils.gitlab.io/modern-cmake/chapters/testing/googletest.html
